### PR TITLE
ARROW-8847: [C++] Pass task hints in Executor API

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Review by lidavidm — APPROVED** [review]
LGTM. Thanks for looking at this :) This could be reasonably easily extended to cover [ARROW-8667](https://issues.apache.org/jira/browse/ARROW-8667) as well, if we do want to have a separate consumer ID hint (and some way to generate an unused ID).

**Comment by pitrou** [comment]
That requires passing the "consumer ID" somehow. Perhaps in the AsyncContext?

**Comment by lidavidm** [comment]
> That requires passing the "consumer ID" somehow. Perhaps in the AsyncContext?

Yeah, that would make sense.

File: cpp/src/arrow/io/interfaces.h [hunk 1]
```diff
@@ -51,6 +51,8 @@ struct ReadRange {
 // EXPERIMENTAL
 struct ARROW_EXPORT AsyncContext {
   ::arrow::internal::Executor* executor;
+  // An application-specific ID, forwarded to executor task submissions
+  int64_t external_id = -1;
```
Comment (pitrou) [inline cpp/src/arrow/io/interfaces.h:55]: @lidavidm Is `int64_t` sufficient here?

**Review by fsaintjacques — APPROVED** [review]

File: cpp/src/arrow/io/interfaces.h (hunk 1 above)
Comment (fsaintjacques) [reply cpp/src/arrow/io/interfaces.h:55]: It should, the other option is to pass a `void *`.

**Pull request closed without merging**